src/main/java/hudson/plugins/copyartifact/CopyArtifact.java

```diff
@@ -512,33 +522,131 @@ private static ItemGroup getItemGroup(Run<?, ?> build) {
-            FilePath baseTargetDir, Copier copier, PrintStream console)
-            throws IOException, InterruptedException {
-        FilePath srcDir = selector.getSourceDirectory(src, console);
+    private boolean perform(Run src, Run<?,?> dst, String expandedFilter, @CheckForNull String expandedExcludes, FilePath targetDir, TaskListener listener) throws IOException, InterruptedException {
+        PrintStream console = listener.getLogger();
+        VirtualFile srcDir = selector.getArtifacts(src, console);
         if (srcDir == null) {
             return isOptional();  // Fail build unless copy is optional
         }
-
-        copier.initialize(src, dst, srcDir, baseTargetDir);
+        Map<String, String> fingerprints = null; // entry → MD5
         try {
-            int cnt;
-            if (!isFlatten())
-                cnt = copier.copyAll(srcDir, expandedFilter, expandedExcludes, targetDir, isFingerprintArtifacts());
-            else {
-                targetDir.mkdirs();  // Create target if needed
-                FilePath[] list = srcDir.list(expandedFilter, expandedExcludes, false);
-                for (FilePath file : list)
-                    copier.copyOne(file, new FilePath(targetDir, file.getName()), isFingerprintArtifacts());
-                cnt = list.length;
+            if (targetDir.isRemote()) {
+                VirtualFile srcDirRemote = srcDir.asRemotable();
+                if (srcDirRemote != null) {
+                    // Perform listing and copying on the agent side so we do not need to stream data from the master.
+                    fingerprints = targetDir.act(new Copy(targetDir, srcDirRemote, expandedFilter, expandedExcludes, isFingerprintArtifacts(), listener, isFlatten()));
+                } else {
+                    fingerprints = new Copy(targetDir, srcDir, expandedFilter, expandedExcludes, isFingerprintArtifacts(), listener, isFlatten()).invoke(null, null);
+                }
+            } else {
+                fingerprints = new Copy(targetDir, srcDir, expandedFilter, expandedExcludes, isFingerprintArtifacts(), listener, isFlatten()).invoke(null, null);
             }
-
+            int cnt = fingerprints.size();
             console.println(Messages.CopyArtifact_Copied(cnt, HyperlinkNote.encodeTo('/'+ src.getParent().getUrl(), src.getParent().getFullDisplayName()),
                     HyperlinkNote.encodeTo('/'+src.getUrl(), Integer.toString(src.getNumber()))));
             // Fail build if 0 files copied unless copy is optional
             return cnt > 0 || isOptional();
         } finally {
-            copier.end();
+            if (fingerprints != null) {
+                Map<String, String> fingerprintsShallow = new HashMap<>();
+                FingerprintMap map = Jenkins.get().getFingerprintMap();
+                for (Map.Entry<String, String> entry : fingerprints.entrySet()) {
+                    String name = entry.getKey().replaceFirst(".+/", "");
```

Comment:
Note for me: snipping ".+/" is equivalent to the current implementation calling `FilePath#getName()` in `FingerprintingCopyMethod`.
(I'm not sure this is a good behavior as reported in [JENKINS-16987](https://issues.jenkins-ci.org/browse/JENKINS-16987))